src/messaging/Channel.h

```diff
@@ -0,0 +1,248 @@
+ */
+
+#pragma once
+
+#include <inet/IPAddress.h>
+#include <transport/raw/MessageHeader.h>
+
+namespace chip {
+namespace Messaging {
+
+/**
+ *  @brief
+ *    The ChannelBuilder object provides information to build a Channel.
+ *
+ *    ChannelBuilder can be used by application to provide information to
+ *    request a channel to a peer. When a channel is requested via
+ *    ExchangeManager::EstablishChannel, a ChannelHandle will be return to
+ *    represent the state of the channel
+ *
+ *    The ChannelBuilder object should be short-live and it is only used within
+ *    ExchangeManager::EstablishChannel call, after the call its state will be
+ *    copied into an internal object represented by ChannelHandle, then the
+ *    ChannelBuilder can be safely released.
+ */
+class ChannelBuilder
+{
+public:
+    enum class NetworkPreference
+    {
+        kNetwork_Default,
+        kNetwork_HighBandwidth,
+        kNetwork_LowLatency,
+        kNetwork_LowPowerConsumption,
+    };
+
+    enum class TransportPreference
+    {
+        kTransport_Connectionless,
+        kTransport_PreferConnectionOriented, // will fallback to connectionless TCP is not supported
+        kTransport_ConnectionOriented,       // will fail if TCP is not supported
+
+        kTransport_Default = kTransport_Connectionless,
+    };
+
+    enum class SessionType
+    {
+        kSession_PASE, // Use SPAKE2 key exchange
+        kSession_CASE, // Use SIGMA key exchange
+
+        kSession_Default = kSession_CASE,
+    };
+
+    ChannelBuilder & SetPeerNodeId(NodeId peerNodeId)
+    {
+        mPeerNodeId = peerNodeId;
+        return *this;
+    }
+    NodeId GetPeerNodeId() const { return mPeerNodeId; }
+
+    ChannelBuilder & SetNetworkPreference(NetworkPreference preference)
+    {
+        mNetworkPreference = preference;
+        return *this;
+    }
+    NetworkPreference GetNetworkPreference() const { return mNetworkPreference; }
+
+    ChannelBuilder & SetTransportPreference(TransportPreference preference)
+    {
+        mTransportPreference = preference;
+        return *this;
+    }
+    TransportPreference GetTransportPreference() const { return mTransportPreference; }
+
+    ChannelBuilder & SetSessionType(SessionType preference)
+    {
+        mSessionType = preference;
+        return *this;
+    }
+    SessionType GetSessionType() const { return mSessionType; }
+
+    uint16_t GetPeerKeyID() const { return mCaseParameters.mPeerKeyId; }
+    ChannelBuilder & SetPeerKeyID(uint16_t keyId)
+    {
+        if (mSessionType == SessionType::kSession_CASE)
+            mCaseParameters.mPeerKeyId = keyId;
+        return *this;
+    }
+
+    uint32_t GetPeerSetUpPINCode() const { return mPaseParameters.mPeerSetUpPINCode; }
+    ChannelBuilder & SetPeerSetUpPINCode(uint32_t peerPINCode)
+    {
+        if (mSessionType == SessionType::kSession_PASE)
+            mPaseParameters.mPeerSetUpPINCode = peerPINCode;
+        return *this;
+    }
+
+#ifndef NDEBUG
```

Comment:
why do we provide this in debug builds?

Please add a comment. I would also not name it 'Hint' but rather `ForcePeerAddress` since the implementation seems to say that it skips resolution if this is provided.

Reply:
At the time writing this PR, the mDNS client is not ready yet. And lots of examples and testings are using addresses directly. So in practice, we need to support this scenario for a while to maintain compatibility and make CI happy.